src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,99 @@
 
 import com.endava.internship.domain.Privilege;
 import com.endava.internship.domain.User;
-import com.endava.internship.service.UserService;
 
+import java.util.AbstractMap;
+import java.util.Arrays;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                map(User::getFirstName).
+                sorted(Comparator.reverseOrder()).
+                collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                sorted(Comparator.comparing(User::getFirstName)).
+                sorted(Comparator.comparing(User::getAge).reversed())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                map(User::getPrivileges).
+                flatMap(List::stream).
+                collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                filter(user -> user.getAge() > age && user.getPrivileges().contains(Privilege.UPDATE))
```

Comment:
Why do we need to check privelege here

Reply:
/**
     * Should return optional of user that has UPDATE privilege
     * and age higher that given age
     *
     * @param users list of users
     * @param age   conditional age
     * @return optional of found user or optional empty if user wasn't found
     */
    Optional<User> getUpdateUserWithAgeHigherThan(List<User> users, int age);

Reply:
all good